### ISSUE TITLE
Question: how to track down what is clearing the cache on shutdown...

### ISSUE THREAD

**Issue by allen-servedio** [ben-manes/caffeine#145]
I read issue #104 and implemented it as suggested as I would like my removalListener to run both as items are evicted from the cache and to run for the remaining items in the cache as the service is shutting down (basically dump the cache when it is doing an orderly shutdown).

However, while I can see before shutting down the service that there are values in the cache, when my shutdown hook runs the cache is completely empty (I call "asMap" on the cache to determine this) AND the removalListener code has not run yet.

I verified that the only two user land shutdown hooks that have been registered are mine and the java.util.logging.LoggingManager.Cleaner class.

As described in #104 I have the executor service that is used by the cache, I call shutdown() on it and I awaitTermination (doing a shutdownNow if that times out). So even if something evicted them before my hook is called, I would think that they would be processed (esp. since I call invalidateAll() on the Cache and then call cleanup() on the cache - however, I don't see anything in the cache at that point).

Without seeing code, I know that you are going to be kinda limited in what you could do, but do you have any suggestions on where I should be looking to see what is going on? Since this is during shutdown, debugging is stone age stuff (like println) - but any suggestions are welcome.

Thanks!

**Comment by ben-manes** [comment]
If you are using the default executor, `ForkJoinPool.commonPool()`, it does not respect shutdown. So that could cause some confusion.

An alternative to `RemovalListener` is `CacheWriter#delete`, which is called synchronously with the operation. This imposes the operation's latency on a calling thread, which could be a negative if user-facing. That would simplify your case, though, so worse case it might be a good stop gap to iterate on this discussion.

In #104, they wanted to close connections gracefully. When you say "dump the cache when it is doing an orderly shutdown" it sounds like you might be trying to allow for a warm restart. In that case you may prefer to use `Policy`, e.g. obtain the top-k entries in the cache. That's a bit of a reach, but thought I'd mention it if perhaps appropriate.

Otherwise maybe you can put together a unit test for us to debug against?

**Comment by ben-manes** [comment]
Please also ensure that you are using the current release (2.4). It fixed some two races causing stale notification values under a high write load. Doubtfully your issue, so hopefully you're able to find the time to dig into this.

**Comment by ben-manes** [comment]
@allen-servedio Do you think that you will be able to look into this soon?

**Comment by allen-servedio** [comment]
Sorry Ben. I should not have opened this without more time set aside to look into it. The current project I am on is taking more time than I anticipated. I will try to block some time this week to try out what you suggested and report back (if it worked) or provide a unit test (if it did not). Thank you for your patience.

**Comment by allen-servedio** [comment]
Hi,

Okay, so I am not yet able to reproduce this in a unit test (I pasted the code that I tried below). The code below is roughly what I am doing in my code when interacting with the cache. The difference between what I see and what this test shows is that when the shutdown hook code runs, the cache is 100% empty already AND the data that was in it was not processed by my removalListener code. 

I was able to determine this via the follow code (you will see it commented out in the larger script below.

```
  val cacheAsMap = cache.asMap()
  println(s"SHUTDOWN CACHE MAP: $cacheAsMap")
  cacheAsMap.asScala.foreach{
    case (key, value) =>
      println(s"$key = $value")
      cacheAsMap.remove(key)
  }
```

The above code prints an empty map in the first println in my code... In the script below, it prints out the cache as you would expect.

Here is the result of running the script below:

```
$ amm evict-on-shutdown-example.sc
Incremented first
Incremented second
Incremented second
Incremented third
Incremented third
Incremented third
first = 1 because EXPLICIT
third = 3 because EXPLICIT
second = 2 because EXPLICIT
Did not have to force shutdown.
```

So, that works 100% as expected.

I have a bad feeling that this started failing when I dragged in some dependency that is doing something clever... **Do you want me to close this ticket until I have something more concrete? It could take me some time before I find something. I could reopen it or comment if I find something useful.**

Here is the script:
```
#!/usr/bin/env amm

// Uses https://github.com/lihaoyi/ammonite 0.8.5 - so it has to be installed to run this.
// After install, run with: amm evict-on-shutdown-example.sc

import $ivy.`com.github.ben-manes.caffeine:caffeine:2.4.0`, com.github.benmanes.caffeine.cache._
import java.util.concurrent.{Executors, TimeUnit}
import java.util.concurrent.atomic.LongAdder
import scala.util.{Failure, Success, Try}
import $ivy.`org.scala-lang.modules::scala-java8-compat:0.8.0`, scala.compat.java8.FunctionConverters._
import scala.collection.JavaConverters._

val processMessageExecutorService =
  Executors.newWorkStealingPool(Runtime.getRuntime.availableProcessors)

val cache: Cache[String, LongAdder] = Caffeine.newBuilder()
  .maximumSize(1000)
  .expireAfterWrite(60, TimeUnit.SECONDS)
  .removalListener(new RemovalListener[String, LongAdder]() {
    override def onRemoval(key: String, value: LongAdder, cause: RemovalCause): Unit = {
      println(s"$key = $value because $cause")
    }
  })
  .executor(processMessageExecutorService)
  .build[String, LongAdder]()

def increment(key: String) = {
  val wasIncremented = Try(cache.get(key, ((_: String) => new LongAdder()).asJava).increment())
  wasIncremented match {
    case Success(_) => println(s"Incremented $key")
    case Failure(ex) => println(s"Failed to increment $key: $ex")
  }
}

sys.addShutdownHook {
  // val cacheAsMap = cache.asMap()
  // println(s"SHUTDOWN CACHE MAP: $cacheAsMap")
  // cacheAsMap.asScala.foreach{
  //   case (key, value) =>
  //     println(s"$key = $value")
  //     cacheAsMap.remove(key)
  // }

  cache.invalidateAll()
  cache.cleanUp()
  processMessageExecutorService.shutdown()

  Try(processMessageExecutorService.awaitTermination(60, TimeUnit.SECONDS)) match {
    case Success(terminated) => if (!terminated) processMessageExecutorService.shutdownNow() else println("Did not have to force shutdown.")
    case Failure(ex) =>
      println(s"Failure during process message termination: $ex")
      processMessageExecutorService.shutdownNow()
  }
}

increment("first")
increment("second")
increment("second")
increment("third")
increment("third")
increment("third")
```

**Comment by ben-manes** [comment]
Just to verify, you are specifying an executor in your production code? And that executor honors `awaitTermination` (unlike the common FJP)? And that executor indicates it has no pending work? If using a non-standard executor, can you please decorate it to log all incoming tasks in case its dropping them? (e.g. FJP had a bug in early versions of JDK8)

Did using a `CacheWriter` show different (correct) behavior in your production scenario?

Can you try calling these methods concurrently? You can use [ConcurrentTestHarness](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/testing/ConcurrentTestHarness.java). It could be a race condition, which your test isn't verifying.

On that note, you may also need to be careful to ensure calls coming in after shutdown are handled at the call site. That could be done by a volatile shutdown flag, for example.

I'm fine closing or leaving open. My main concern is resolving your issue. If you won't be able to work on this for a while then we could close. Hopefully that's not necessary.

**Comment by allen-servedio** [comment]
Yes on the executor - the line of code you see is pretty close to what is in my code. It calls this method in java.util.concurrent.Executors:

```
    public static ExecutorService newWorkStealingPool() {
        return new ForkJoinPool
            (Runtime.getRuntime().availableProcessors(),
             ForkJoinPool.defaultForkJoinWorkerThreadFactory,
             null, true);
    }
```

This is the version of Java I am running with:

```
$ java -version
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)
```

Because the map is empty by the time the shutdown hook runs, I don't have anything to send to the CacheWriter.delete method...

For shutdown processing of incoming requests, I wind up calling this after I have stopped all incoming requests and they have finished processing (I am able to determine that).

Okay, in thinking about what you are saying around the executor, I take it that you suspect that the cache was actually flushed prior to it hitting the shutdown hook (because I check the contents of the cache right at the start of the shutdown hook code and it is empty already)? And that work is actually pending. But, because of the executor I am using, I don't wait and just shutdown?

Could be... I did try adding a couple of Thread.sleep(5) type of statements in there as a course grained way to let stuff not in my thread to finish up - it did not help.

**Comment by ben-manes** [comment]
The idea behind using `CacheWriter` is that the executor is no longer involved. If the cache is empty then the `delete` method should have been called inline for the previously removed entries. That way there is no asynchronous surprises due to a race in my or your code. If we can verify that works, then we might have an easier time narrowing down the problem.

**Comment by allen-servedio** [comment]
Okay I am closing this issue as user error. I managed to make TWO copies of the cache in two separate places of the app (there was JUST enough indirection that I did not realize the two were unconnected). My shutdown hook referenced the first one (which never got data) and the rest of the code referenced the second one. 

Apologies for wasting your time - but sincerely appreciate your feedback.
